### PR TITLE
COSX Grid Handling Improvements

### PR DESCRIPTION
**Pull request by davpoolechem**
## Description

This PR makes a couple of changes to the handling of grids, and debugging thereof, within the COSX code:

1. The code is refactored so that grids, instead of being stored in separate variables, are combined together in an `unordered_map` structure. Variables reliant on specific grids, such as `Q_mat_`, are treated similarly; and the code is adjusted to match. This allows for easy extension of the COSX code to support more grids in the future, if desired; and allows for smoother implementation of other features in this PR.

2. Control over the active COSX grid now relies on the `current_grid_` string variable instead of the boolean `early_screening_` variable. This fits better with the `unordered_map` arrangement this code proposes, and is not limited to a two-grid scheme.

3. Upon user request (i.e., setting `DEBUG` != 0), information about each grid is printed out. Information printed out includes: total number of grid points, total number of grid blocks, average grid points per block, and average grid points per atom.

4. The timer.dat file presents the amount of time spent in COSX per grid, as well as the total COSX time.

## User API & Changelog headlines
- [X] Setting `DEBUG` to != 0 now prints out specific information on each grid used during a COSX calculation.

## Dev notes & details
- [X] COSX grids, and related variables, are now grouped together in `unordered_map` objects, and the code is adjusted to match.
- [X] Control over active grids is now controlled by `gridopt_` instead of `early_screening_`.
- [X] For COSX calculations, timer.dat now generates COSX timings per-grid, as well as the overall COSX time.
- [X] Informational printout of COSX grids has been added.

## Questions
- [ ] What do we want to do with `early_screening_`? In this PR, it basically has no use, since its functionality has been replaced by that of `gridopt_`. Early screening as a tactic (i.e., using looser ERI screening thresholds for early SCF iterations) _is_ a thing that could be added in the future for which `early_screening_` could be applicable, but that is not yet in Psi4 as far as I know.

## Checklist
- [X] Tests added for any new features
- [X] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [X] Ready for review
- [ ] Ready for merge

### CODE REVIEW

**Review by jturney — APPROVED** [review]

File: psi4/driver/procrouting/scf_proc/scf_iterator.py [hunk 1]
```diff
@@ -489,7 +493,7 @@ def scf_iterate(self, e_conv=None, d_conv=None):
 
         # Print out the iteration
         core.print_out(
-            "   @%s%s iter %3s: %20.14f   %12.5e   %-11.5e %s\n" %
+            "   @%s%s iter %3s: %20.14f   %12.5e   %-11.5e %s \n" %
```
Comment (susilehtola) [inline psi4/driver/procrouting/scf_proc/scf_iterator.py:496]: why the unnecessary whitespace at the end?

File: psi4/driver/procrouting/scf_proc/scf_iterator.py [hunk 2]
```diff
@@ -267,9 +267,13 @@ def scf_iterate(self, e_conv=None, d_conv=None):
     soscf_enabled = _validate_soscf()
     frac_enabled = _validate_frac()
     efp_enabled = hasattr(self.molecule(), 'EFP')
+    cosx_enabled = "COSX" in core.get_option('SCF', 'SCF_TYPE')
 
     # does the JK algorithm use severe screening approximations for early SCF iterations?
     early_screening = self.jk().get_early_screening()
+    if cosx_enabled:
+        self.jk().set_COSX_grid("Initial")
```
Comment (susilehtola) [inline psi4/driver/procrouting/scf_proc/scf_iterator.py:275]: shouldn't this be dependent on the use of `early_screening`?

File: psi4/driver/procrouting/scf_proc/scf_iterator.py [hunk 3]
```diff
@@ -520,6 +524,10 @@ def scf_iterate(self, e_conv=None, d_conv=None):
                 # make note of the change to early screening; next SCF iteration will be the last
                 early_screening_disabled = True
 
+                # cosx uses the largest grid for its final SCF iteration
+                if cosx_enabled:
+                    self.jk().set_COSX_grid("Final")
```
Comment (susilehtola) [inline psi4/driver/procrouting/scf_proc/scf_iterator.py:529]: note that this only runs `if early_screening`!

File: psi4/src/psi4/libfock/COSK.cc [hunk 4]
```diff
@@ -185,85 +185,87 @@ COSK::COSK(std::shared_ptr<BasisSet> primary, Options& options) : SplitJK(primar
...
+            {"DFT_BASIS_TOLERANCE",   options_.get_double("COSX_BASIS_TOLERANCE")},
+            {"DFT_BS_RADIUS_ALPHA",   1.0},
+            {"DFT_PRUNING_ALPHA",     1.0},
+            {"DFT_BLOCK_MAX_RADIUS",  3.0},
+            {"DFT_WEIGHTS_TOLERANCE", 1e-15},
+        };
+        grid = std::make_shared<DFTGrid>(primary_->molecule(), primary_, grid_int_options, grid_str_options, grid_float_options, options_);
+
+        // Print out warning if grid with negative grid weights is used
+        // Original Nesse COSX formulation does not support negative grid weights
```
Comment (susilehtola) [inline psi4/src/psi4/libfock/COSK.cc:228]: Neese

File: psi4/driver/procrouting/scf_proc/scf_iterator.py (hunk 2 above)
Comment (davpoolechem) [reply psi4/driver/procrouting/scf_proc/scf_iterator.py:275]: What to do with the `early_screening` variable is actually something I wasn't quite sure of with this PR, and I mention this in the Questions section of this PR submission.
 
Disconnecting the `early_screening` variable from control over the COSX grid is an intentional design decision, as there are plans to expand COSX to a three-grid scheme, which is incompatible with `early_screening`'s binary nature as a boolean. I kept it in for now, because "early screening" as a concept can refer to other forms of SCF optimization tricks (e.g., starting early SCF iterations with a looser ERI screening tolerance and tightening the tolerance as the SCF progresses).

But since it's not really being used in this PR, it's probably better to remove it. We wouldn't want dead code, after all.

Comment (davpoolechem) [reply psi4/driver/procrouting/scf_proc/scf_iterator.py:275]: Disconnecting the `early_screening` variable from control over the COSX grid is an intentional design decision in this PR. I plan on eventually adding a three-grid scheme to COSX, where three grids are used over the course of the SCF. This is incompatible with grid control via `early_screening`, given the latter is a boolean.

File: psi4/driver/procrouting/scf_proc/scf_iterator.py (hunk 1 above)
Comment (davpoolechem) [reply psi4/driver/procrouting/scf_proc/scf_iterator.py:496]: I suppose I thought it looked a bit neater this way, but I'll be happy to switch it back!

File: psi4/src/psi4/libfock/COSK.cc (hunk 4 above)
Comment (davpoolechem) [reply psi4/src/psi4/libfock/COSK.cc:228]: Ah, my bad! I will fix this!

File: psi4/driver/procrouting/scf_proc/scf_iterator.py (hunk 3 above)
Comment (davpoolechem) [reply psi4/driver/procrouting/scf_proc/scf_iterator.py:529]: Yes, that is correct. Decoupling `early_screening` from direct COSX grid control was one motivation of this PR, to allow for an eventual three-grid COSX SCF scheme.  

Though you make the fair point that `early_screening` still _does_ influence the COSX grid. Maybe I should make some changes to just... get rid of `early_screening` entirely. I wasn't sure what to do with it, honestly, as I mentioned in the Questions section of my PR submission.

File: psi4/driver/procrouting/scf_proc/scf_iterator.py (hunk 1 above)
Comment (davpoolechem) [reply psi4/driver/procrouting/scf_proc/scf_iterator.py]: All right, reverted!

File: psi4/src/psi4/libfock/COSK.cc (hunk 4 above)
Comment (davpoolechem) [reply psi4/src/psi4/libfock/COSK.cc]: I removed the associated warning entirely, so this issue is fixed by proxy.

File: psi4/src/psi4/libfock/COSK.cc [hunk 5]
```diff
@@ -185,85 +185,87 @@ COSK::COSK(std::shared_ptr<BasisSet> primary, Options& options) : SplitJK(primar
...
+        // which can happen with certain grid configurations
+        // See https://github.com/psi4/psi4/issues/2890
+        auto warning_printed = false;
+        for (const auto &block : grid->blocks()) {
+            const auto w = block->w();
+            for (int ipoint = 0; ipoint < block->npoints(); ++ipoint) {
+                if (w[ipoint] < 0.0) {
+                    std::string warning_message = "  INFO: The definition of the current ";
+                    warning_message += gridname_lowercase;
+                    warning_message += " grid includes negative weights, which the original COSX formulation does not support!\n    If this is of concern, please choose another final grid through adjusting either COSX_PRUNING_SCHEME or COSX_SPHERICAL_POINTS_";
```
[the remark this answers is not in the document]
Comment (davpoolechem) [reply psi4/src/psi4/libfock/COSK.cc]: All right, removed!

File: psi4/src/psi4/libfock/SplitJK.h [hunk 6]
```diff
@@ -293,6 +308,9 @@ class PSI_API COSK : public SplitJK {
     * print name of method
     */
     std::string name() override { return "COSX"; }
+
+    void set_COSX_grid(std::string gridopt) override { gridopt_ = gridopt; };
+    std::string get_COSX_grid() override { return gridopt_; };
```
[the remark this answers is not in the document]
Comment (davpoolechem) [reply psi4/src/psi4/libfock/SplitJK.h]: Done! I renamed `gridopt_` to `current_grid_`, which hopefully is more clear.

File: psi4/src/export_fock.cc [hunk 7]
```diff
@@ -200,7 +200,9 @@ void export_fock(py::module &m) {
 
     py::class_<CompositeJK, std::shared_ptr<CompositeJK>, JK>(m, "CompositeJK", "docstring")
         .def("do_incfock_iter", &CompositeJK::do_incfock_iter, "Was the last Fock build incremental?")
-        .def("clear_D_prev", &CompositeJK::clear_D_prev, "Clear previous D matrices.");
+        .def("clear_D_prev", &CompositeJK::clear_D_prev, "Clear previous D matrices.")
+        .def("set_COSX_grid", &CompositeJK::set_COSX_grid, "Set grid to use for COSX.")
+        .def("get_COSX_grid", &CompositeJK::get_COSX_grid, "Return grid used for COSX.");
```
[the remark this answers is not in the document]
Comment (davpoolechem) [reply psi4/src/export_fock.cc]: Done and done!

File: psi4/src/psi4/libfock/CompositeJK.cc [hunk 8]
```diff
@@ -288,28 +289,38 @@ void CompositeJK::compute_JK() {
...
+        timer_off("CompositeJK: " + j_algo_->name());
     }
 
     // Exchange Matrix
     if (do_K_) {
-        timer_on("CompositeJK: K");
+        timer_on("CompositeJK: " + k_algo_->name());
+
+        if (k_algo_->name() == "COSX") {
+            std::string gridname = early_screening_ ? "Initial" : "Final";
```
[the remark this answers is not in the document]
Comment (davpoolechem) [reply psi4/src/psi4/libfock/CompositeJK.cc]: Ah, nice catch! Yeah, it shouldn't depend on `early_screening_` here. Fixed!

[the remark this answers is not in the document]
Comment (davpoolechem) [reply psi4/src/psi4/libfock/CompositeJK.cc]: Also, in light of this comment, I did some small refactoring to completely remove the `early_screening_` variable C++-side. This was the only point at which `early_screening_` was actually utilized, so now all control of any remaining "early screening" is done through the Python driver.

**Review by loriab — APPROVED** [review]
